[WebGPU] Timestamp writes only work in vary specific situations

litherum opened the pull request:
#### 7db946a6f19e50aa9ff1243097142c2a46a89e43
<pre>
[WebGPU] Timestamp writes only work in vary specific situations
<a href="https://bugs.webkit.org/show_bug.cgi?id=251773">https://bugs.webkit.org/show_bug.cgi?id=251773</a>
rdar://105067419

Reviewed by Tadeu Zagallo.

Before this patch, the infrastructure for timestamp writes is very specific:
- Only Apple Silicon machines are supported
- CommandEncoder.writeTimestamp() is not supported
- ComputePassDescriptor.timestampWrites are not supported
- RenderPassDescriptor.timestampWrites is supported, but only for the first element,
      and only for specific hardcoded values of queryIndex

This patch generalizes support:

In Metal, different devices support different query set APIs. You have to call
[MTLDevice supportsCounterSampling:] to know which kind of device you&apos;re on. Apple
Silicon devices report that they only support counter sampling on stage boundaries,
via the pass descriptors&apos; sampleBufferAttachments. All other devices (that I&apos;ve
tried) report that they support counter sampling on individual command boundaries.
If a device doesn&apos;t support either of these granularities, then it just won&apos;t
support the timestamp queries feature in the first place.

For pass descriptors&apos; timestampWrites array, if the device is an Apple Silicon device,
we are supposed to be able to naively transform the timestampWrites into
sampleBufferAttachments, which is what this patch does. However, rdar://91371495 is
about how that doesn&apos;t actually work; a follow-up patch is necessary to work around
this. For non-Apple-Silicon devices, we implement the timestampWrites array by just
calling -[CommandEncoder sampleCountersInBuffer:] at the beginning and end of the pass.

For CommandEncoder.writeTimestamp(), we can implement this on Apple Silicon devices by
creating a dummy blit pass, and setting sampleBufferAttachments on it. We want to
coalesce multiple writes into a single blit pass, so we defer doing this as long as
possible - until either we would have had to make a blit pass anyway, we have to start
a new pass, or the command encoder is ending. For non-Apple-Silicon devices,
CommandEncoder.writeTimestamp() directly turns into
-[MTLBlitCommandEncoder sampleCountersInBuffer:].

Because of rdar://91371495, this patch isn&apos;t actually fully-functional; however,
the existing infrastructure wasn&apos;t fully-functional either, so I&apos;m kind of viewing this
patch as not-a-regression. My next patch will work around rdar://91371495.

* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::ensureBlitCommandEncoder):
(WebGPU::CommandEncoder::finalizeBlitCommandEncoder):
(WebGPU::CommandEncoder::validateComputePassDescriptor const):
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::writeTimestamp):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
(WebGPU::ComputePassEncoder::create):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::ComputePassEncoder):
(WebGPU::ComputePassEncoder::endPass):
* Source/WebGPU/WebGPU/HardwareCapabilities.h:
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::baseCapabilities):
(WebGPU::mergeBaseCapabilities):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
(WebGPU::RenderPassEncoder::create):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::endPass):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::createVertexDescriptor):

Canonical link: <a href="https://commits.webkit.org/259914@main">https://commits.webkit.org/259914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0f9364bf6f7bafdf9403ab01ec6cefc5c689073

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106295 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15349 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39134 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115488 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16791 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6565 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98511 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115175 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112059 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12776 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95757 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40330 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94651 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27400 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82020 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8589 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28752 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9099 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5323 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14712 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48298 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6849 "'git push ...'") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10635 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -147,11 +204,13 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
         return RenderPassEncoder::createInvalid(m_device);
 
-    MTLRenderPassDescriptor* mtlDescriptor = [MTLRenderPassDescriptor renderPassDescriptor];
+    MTLRenderPassDescriptor* mtlDescriptor = [MTLRenderPassDescriptor new];
```
this change is fine, the two methods are equivalent

litherum replied:
Right. I slightly prefer `new` rather than the other one, because the other one creates an autoreleased object, whereas ARC will release the `new`'ed object at the end of the lexical scope. So the lifetime is more tightly bounded.

But in reality there is probably no measurable difference

mwyrzykowski reviewed (commented):
This breaks query-sets.html, is that sample setup incorrectly?

mwyrzykowski reviewed (changes requested):
Can we either please fix the regression causing regressed behavior in the query-sets.html sample or, if the sample was using the API incorrectly, modify the sample to correctly use the API please?

litherum committed:
[WebGPU] Timestamp writes only work in vary specific situations
https://bugs.webkit.org/show_bug.cgi?id=251773
rdar://105067419

Reviewed by Tadeu Zagallo.

Before this patch, the infrastructure for timestamp writes is very specific:
- Only Apple Silicon machines are supported
- CommandEncoder.writeTimestamp() is not supported
- ComputePassDescriptor.timestampWrites are not supported
- RenderPassDescriptor.timestampWrites is supported, but only for the first element,
      and only for specific hardcoded values of queryIndex

This patch generalizes support:

In Metal, different devices support different query set APIs. You have to call
[MTLDevice supportsCounterSampling:] to know which kind of device you're on. Apple
Silicon devices report that they only support counter sampling on stage boundaries,
via the pass descriptors' sampleBufferAttachments. All other devices (that I've
tried) report that they support counter sampling on individual command boundaries.
If a device doesn't support either of these granularities, then it just won't
support the timestamp queries feature in the first place.

For pass descriptors' timestampWrites array, if the device is an Apple Silicon device,
we are supposed to be able to naively transform the timestampWrites into
sampleBufferAttachments, which is what this patch does. However, rdar://91371495 is
about how that doesn't actually work; a follow-up patch is necessary to work around
this. For non-Apple-Silicon devices, we implement the timestampWrites array by just
calling -[CommandEncoder sampleCountersInBuffer:] at the beginning and end of the pass.

For CommandEncoder.writeTimestamp(), we can implement this on Apple Silicon devices by
creating a dummy blit pass, and setting sampleBufferAttachments on it. We want to
coalesce multiple writes into a single blit pass, so we defer doing this as long as
possible - until either we would have had to make a blit pass anyway, we have to start
a new pass, or the command encoder is ending. For non-Apple-Silicon devices,
CommandEncoder.writeTimestamp() directly turns into
-[MTLBlitCommandEncoder sampleCountersInBuffer:].

Because of rdar://91371495, this patch isn't actually fully-functional; however,
the existing infrastructure wasn't fully-functional either, so I'm kind of viewing this
patch as not-a-regression. My next patch will work around rdar://91371495.

* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::ensureBlitCommandEncoder):
(WebGPU::CommandEncoder::finalizeBlitCommandEncoder):
(WebGPU::CommandEncoder::validateComputePassDescriptor const):
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::writeTimestamp):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
(WebGPU::ComputePassEncoder::create):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::ComputePassEncoder):
(WebGPU::ComputePassEncoder::endPass):
* Source/WebGPU/WebGPU/HardwareCapabilities.h:
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::baseCapabilities):
(WebGPU::mergeBaseCapabilities):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
(WebGPU::RenderPassEncoder::create):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::endPass):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::createVertexDescriptor):

Canonical link: https://commits.webkit.org/259914@main
